src/pd/client.rs

```diff
@@ -11,247 +11,408 @@
+
+        let client = match connect(ep) {
+            Ok(c) => c,
+            // Ignore failed PD node.
+            Err(e) => {
+                error!("PD endpoint {} is down: {:?}", ep, e);
+                continue;
+            }
+        };
+
+        let resp = match client.GetMembers(pdpb::GetMembersRequest::new()) {
+            Ok(resp) => resp,
+            // Ignore failed PD node.
+            Err(e) => {
+                error!("PD endpoint {} failed to respond: {:?}", ep, e);
+                continue;
+            }
+        };
+
+        // Check cluster ID.
+        let cid = resp.get_header().get_cluster_id();
+        if let Some(sample) = cluster_id {
+            if sample != cid {
+                return Err(box_err!("PD response cluster_id mismatch, want {}, got {}",
+                                    sample,
+                                    cid));
+            }
+        } else {
+            cluster_id = Some(cid);
+        }
+        // TODO: check all fields later?
 
-    try!(stream.set_read_timeout(Some(Duration::from_secs(SOCKET_READ_TIMEOUT))));
-    let mut resp = Message::new();
-    let id = try!(rpc::decode_msg(stream, &mut resp));
-    if resp.get_msg_type() != MessageType::PdResp {
-        return Err(box_err!("invalid pd response type {:?}", resp.get_msg_type()));
+        if members.is_none() {
+            members = Some(resp);
+        }
     }
-    timer.observe_duration();
 
-    Ok((id, resp.take_pd_resp()))
+    info!("All PD endpoints are consistent, {:?}", endpoints);
+
+    match members {
+        Some(members) => {
+            let (client, members) = box_try!(try_connect_leader(&members));
+            Ok((client, members))
+        }
+        _ => Err(box_err!("PD cluster failed to respond")),
+    }
 }
 
-fn rpc_connect(endpoint: &str) -> Result<TcpStream> {
-    let mut stream = try!(make_std_tcp_conn(endpoint));
-    try!(stream.set_write_timeout(Some(Duration::from_secs(SOCKET_WRITE_TIMEOUT))));
+fn connect(addr: &str) -> Result<pdpb_grpc::PDClient> {
+    debug!("connect to PD endpoint: {:?}", addr);
+    let (host, port) = match Url::parse(addr) {
+        Ok(ep) => {
+            let host = match ep.host_str() {
+                Some(h) => h.to_owned(),
+                None => return Err(box_err!("unkown host, please specify the host")),
+            };
+            let port = match ep.port() {
+                Some(p) => p,
+                None => return Err(box_err!("unkown port, please specify the port")),
+            };
+            (host, port)
+        }
 
-    // Send a HTTP header to tell PD to hijack this connection for RPC.
-    let header_str = format!("GET {} HTTP/1.0\r\n\r\n", PD_RPC_PREFIX);
-    let header = header_str.as_bytes();
-    match stream.write_all(header) {
-        Ok(_) => Ok(stream),
-        Err(err) => Err(box_err!("failed to connect to {} error: {:?}", endpoint, err)),
+        Err(_) => {
+            let mut parts = addr.split(':');
```

Comment:
For the very first time, `addr` is passed from command line, eg, `127.0.0.1:2379`, `Url::parse()` returns an error.